### ISSUE TITLE
Samples, switch themes, and save don't mix

### ISSUE THREAD

**Issue by aerwin** [maqetta/maqetta#3111]
- Open Sample1.html
- Choose Switch theme
- Get dialog that says file has unsaved changes and click the save button. See errors in console, but not problems on screen:

```
PUT http://maqetta-app-test.austin.ibm.com/maqetta/project1/samples/Sample1.html 404 (Not Found) dojo.js:4720
_4fc
 dojo.js:5355
An error occurred: davinci.model.resource.File.prototype.setContents RequestError: Unable to load ./project1/samples/Sample1.html status: 404 : ./project1/samples/Sample1.html 
```
- Move forward changing theme to Sketch and see rendering properly update
- Try to save Sample1.html and get dialog that need to pick a writable location. Choose root directory and click ok. See another error in console:

```
Uncaught TypeError: Cannot call method 'forEach' of undefined davinci.js:27893
_13f3.changeThemeBase davinci.js:27893
_13f3.rebuildSource davinci.js:27820
_23f4.saveAs._2423 davinci.js:50535
(anonymous function) davinci.js:62019
lang.hitch dojo.js:4960
lang.hitch dojo.js:4960
_55f.(anonymous function)._564 dojo.js:6939
_1c1e._onSubmit davinci.js:39216
_23d0._onClick davinci.js:50332
_272 dojo.js:2795
_e98._onClick davinci.js:20911
lang.hitch dojo.js:4956
_d4 dijit.js:674
```
- Close page
- Reopen from new location and page is blank

Source at end of scenario has no widgets in it

```
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8"/>
<title>Untitled</title>
<script type="text/javascript" src="lib/dojo/dojo/dojo.js" data-dojo-config="'parseOnLoad':true,'packages':[{'name':'gridx','location':'../gridx'},{'name':'clipart','location':'../../clipart'},{'name':'maqettaSamples','location':'../../../samples'},{'name':'maqetta','location':'../../maqetta'},{'name':'shapes','location':'../../shapes'},{'name':'zazl','location':'../../zazl'},{'name':'widgets','location':'../../custom'}]"></script>
<script type="text/javascript">
require([
  "dijit/dijit",
  "dojo/parser",
  "maqetta/space",
  "maqetta/AppStates"
]);
</script>
<style>@import "themes/claro/document.css";@import "themes/claro/claro.css";
</style>
</head>
<body class="claro" data-maq-flow-layout="true" data-maq-ws="collapse" id="myapp" data-maq-appstates="{}">
</body>
</html>

```

**Comment by jhpedemonte** [comment]
Who last looked into the "save for read-only files" bug?  Was it @doronrosenberg or @childsb?

@JonFerraiolo

**Comment by jhpedemonte** [comment]
-> @doronrosenberg, waiting for answer.

**Comment by doronrosenberg** [comment]
The issue here is that the changes in file, save before applying new theme dialog is not doing readonly checks.

**Comment by doronrosenberg** [comment]
@JonFerraiolo @jhpedemonte 

So this is an interesting issue.  The initial problem was that Samples1.html was being marked as dirty when opened, which I found a fix for.

This lead me down a path to another issue - modifying the theme of a readonly file that is dirty is broken.

For a read only file, the dialog that pops up asking if you want to save/cancel after clicking switch theme won't work for readonly as we need to do a saveas first.  But if we do a saveas, we have this async save and loading up of a new editor that has to happen first before we can continue doing the theme modifying action.

Sounds like a dangerous fix for M7 to me.  We also have no callback mechanism for saveas which makes it much harder.  We need a better indicator that you have a readonly file opened and then disable certain actions.

Thoughts?

**Comment by JonFerraiolo** [comment]
@peller @billreed63 @childsb 

> Sounds like a dangerous fix for M7 to me

Yes indeed. I suggest pushing this to M8, and then address this issue as part of a more general attack on dealing with read-only files. 

@childsb has looked at this issue in the past.

I suspect that implementing suitable protection against performing operations on read-only files (and supplemental resources such as CSS files) will represent a multi-day programming effort. My thoughts about how to attack this problem would be to try to put checks into the various `run()` functions within our various "action" classes so that we can stop various operations before they even get started. If a given operation could not be completed successfully due to read-only files, then we would post a dialog that tells the user what they need to do in order to be able to perform the given operation successfully. (In some cases, this would involve doing a SaveAs operation into a read-write folder, and then trying the operation again.)

**Comment by jhpedemonte** [comment]
-> M8, too risky now.

**Comment by JonFerraiolo** [comment]
-> M8 Prio-Med

Note that I added logic to NewFile.html to ensure that the new file isn't in a read-only folder. Not exactly the same thing as here, but might be useful as a reference: See 15fa837dcdd03ad547810c84891431ab9bf8f795

Maybe what we need to do is supplement the "save" command and "save as" logic to similarly check for read-only status. If the file is read-only and the user attempts to "save" as the file is read-only, then we switch to a save-as operation. And with the save-as operation, if the current folder is read-only, then switch to the root folder as the current folder.

Another possible wrinkle is that if opening a read-only file, we could change the filename to something like "Sample1 copy.html' immediately upon opening the editor, and then immediately change the current folder to root folder if current folder if read-only.
